### ISSUE TITLE
Unit Tests

### ISSUE THREAD

**Issue by dilijev** [wxMaxima-developers/wxmaxima#258]
It might be a good idea to add unit tests to the project, to test corner cases and ensure that nothing breaks when new features are added or bugs are fixed.

Since the majority of the application is GUI, I can see why this might not make sense, but while I was testing a fix for another issue #213, I created a function whose functionality didn't depend on state, so I could make it a static function, and perform unit tests as a faster way to debug the issue.

As a result of this I have code for some unit tests, and manually created a target for unit testing, but the problem is that, at present, I don't know how automake works or how to add new targets to it.

Here's my code:

wxMaximaTest.cpp

``` C++
#include <stdexcept>
#include <cstdlib>                            // srand
#include <string>
#include <iostream>
#include <sstream>

#include "cppunit/extensions/HelperMacros.h"  // CPPUNIT_TEST, CPPUNIT_TEST_SUITE, CPPUNIT_TEST_SUITE_END
#include "cppunit/TestFixture.h"              // TestFixture
#include "cppunit/TextTestRunner.h"           // TestRunner

#include <wx/string.h>

#include "./wxMaxima.h"

using namespace std;

#define ASSERT(b) CPPUNIT_ASSERT(b)

struct TestStripComments : CppUnit::TestFixture {
    void test_strip_comments_0() {
        wxString s = wxT(";\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT(";\n"));
    }

    void test_strip_comments_1() {
        wxString s = wxT("/*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT(";\n"));
    }

    void test_strip_comments_2() {
        wxString s = wxT(" /*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT(";\n"));
    }

    void test_strip_comments_3() {
        wxString s = wxT(";/*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT(";\n"));
    }

    void test_strip_comments_4() {
        wxString s = wxT("; /*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT(";\n"));
    }

    void test_strip_comments_5() {
        wxString s = wxT("x=1;/*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT("x=1;\n"));
    }

    void test_strip_comments_6() {
        wxString s = wxT("x=1; /*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT("x=1;\n"));
    }

    void test_strip_comments_7() {
        wxString s = wxT("x=1;;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT("x=1;\n"));
    }

    void test_strip_comments_8() {
        wxString s = wxT("x=1; /*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT("x=1;\n"));
    }

    void test_strip_comments_9() {
        wxString s = wxT("x=1; /*test*/;;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT("x=1;\n"));
    }

    void test_strip_comments_10() {
        wxString s = wxT("x=1; /*test*/;; /*test*/;\n");
        wxMaxima::StripComments(s);
        cout << s.ToAscii() << endl;

        ASSERT(s == wxT("x=1;\n"));
    }

    // -----
    // suite
    // -----

    CPPUNIT_TEST_SUITE(TestStripComments);
        CPPUNIT_TEST(test_strip_comments_0);
        CPPUNIT_TEST(test_strip_comments_1);
        CPPUNIT_TEST(test_strip_comments_2);
        CPPUNIT_TEST(test_strip_comments_3);
        CPPUNIT_TEST(test_strip_comments_4);
        CPPUNIT_TEST(test_strip_comments_5);
        CPPUNIT_TEST(test_strip_comments_6);
        CPPUNIT_TEST(test_strip_comments_7);
        CPPUNIT_TEST(test_strip_comments_8);
        CPPUNIT_TEST(test_strip_comments_9);
        CPPUNIT_TEST(test_strip_comments_10);
    CPPUNIT_TEST_SUITE_END();
};

// ----
// main
// ----

int main() {
    using namespace std;

    try {
        ios_base::sync_with_stdio(false);  // turn off sync with C I/O
        cout << "wxMaximaTest.c++" << endl;

        CppUnit::TextTestRunner tr;
        tr.addTest(TestStripComments::suite());
        tr.run();
    } catch (exception& e) {
        cout << e.what() << endl;
    }

    cout << "Done." << endl;
    return 0;
}
```

I added a different set of compile objects, where I replaced the main.o with the wxMaximaTest.o, which contains the main for the unit tests.

```
wxmaxima_OBJECTS_TEST = Config.$(OBJEXT) wxMaximaTest.$(OBJEXT) \
    wxMaxima.$(OBJEXT) wxMaximaFrame.$(OBJEXT) \
    SubstituteWiz.$(OBJEXT) IntegrateWiz.$(OBJEXT) \
    LimitWiz.$(OBJEXT) Plot2dWiz.$(OBJEXT) SeriesWiz.$(OBJEXT) \
    SumWiz.$(OBJEXT) Plot3dWiz.$(OBJEXT) Gen1Wiz.$(OBJEXT) \
    Gen2Wiz.$(OBJEXT) Gen3Wiz.$(OBJEXT) Gen4Wiz.$(OBJEXT) \
    BC2Wiz.$(OBJEXT) SystemWiz.$(OBJEXT) BTextCtrl.$(OBJEXT) \
    MatWiz.$(OBJEXT) ExptCell.$(OBJEXT) FracCell.$(OBJEXT) \
    SqrtCell.$(OBJEXT) MatrCell.$(OBJEXT) MathCell.$(OBJEXT) \
    SubCell.$(OBJEXT) IntCell.$(OBJEXT) TextCell.$(OBJEXT) \
    LimitCell.$(OBJEXT) ParenCell.$(OBJEXT) SumCell.$(OBJEXT) \
    AbsCell.$(OBJEXT) AtCell.$(OBJEXT) DiffCell.$(OBJEXT) \
    FunCell.$(OBJEXT) MathCtrl.$(OBJEXT) CellParser.$(OBJEXT) \
    MathParser.$(OBJEXT) MathPrintout.$(OBJEXT) Bitmap.$(OBJEXT) \
    MyTipProvider.$(OBJEXT) EditorCell.$(OBJEXT) ImgCell.$(OBJEXT) \
    SubSupCell.$(OBJEXT) SlideShowCell.$(OBJEXT) \
    GroupCell.$(OBJEXT) EvaluationQueue.$(OBJEXT) \
    History.$(OBJEXT) Autocomplete.$(OBJEXT) \
    PlotFormatWiz.$(OBJEXT)
```

I created a copy of the wxmaxima target to make wxmaximaTest (included the necessary link commands for building on Linux

```
wxmaxima$(EXEEXT): $(wxmaxima_OBJECTS) $(wxmaxima_DEPENDENCIES) 
    @rm -f wxmaxima$(EXEEXT)
    $(wxmaxima_LINK) $(wxmaxima_OBJECTS) $(wxmaxima_LDADD) $(LIBS)

wxmaximaTest$(EXEEXT): $(wxmaxima_OBJECTS_TEST) $(wxmaxima_DEPENDENCIES) 
    @rm -f wxmaximaTest$(EXEEXT)
    $(wxmaxima_LINK) -lcppunit -ldl $(wxmaxima_OBJECTS_TEST) $(wxmaxima_LDADD) $(LIBS)
```

All of this is very specific to my machine and would need to be integrated into the automatic generation process if unit tests were to be added to the project.

**Comment by gunterkoenigsmann** [comment]
Unit tests always are great. Also testpilot and similar might provide a way to test the running program. Did add a .wxmx file to the tarball that at least allows manual tests.

**Comment by dilijev** [comment]
Most of the problems I've had with creating unit tests is that the code that would be interesting to unit test is hidden in many layers of abstraction and GUI code. We need some sort of test bench with mock objects to wrap the relevant code and test the functionality.

**Comment by gunterkoenigsmann** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

We could still test the list management stuff. Even if I have to admit
there isn't much code invoved here. Even if that would mean we have to
make many private or protected methods public so the test bench can
access them. In theory we could #ifndef the private: and protected:
labels during compilation of the testbench to avoid this being
necessary for the real program. The same is true for many helper
routines like my markdown parser.

I assume the main problem with unit testing is that wxWidgets by
making all the interesting stuff a derivate of a gui element makes
programs be built on a structure that has many advantages - but that
at the same time is not built with unit-tests in mind. Restructuring
the the whole program so big parts of it would be unit-test friendly
would be a task that I personally would judge to exceed the man-power
of the community.
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1

iQJ8BAEBCgBmBQJU4Yd3XxSAAAAAAC4AKGlzc3Vlci1mcHJAbm90YXRpb25zLm9w
ZW5wZ3AuZmlmdGhob3JzZW1hbi5uZXQ1M0YwNDdDRTY2QjkxQjBGNzI0QzU0NUQ1
Qzg2QzBFNDIxMUQ1QjhFAAoJEFyGwOQhHVuOWvAP/RoPYiCv00JYklezSRmj8Xcf
xXyktu2p5lR5hyp6+hqnl42Ot9OIqhHvlkQdrdVp2mAcD/SeegMqqlzW/xDqCOcl
5AETulPfejcsl0CJiX4FNxmW8yCSMU9ghsifod3s05Aub0+DQ4YMApk+JXhaNYXP
h3xswGPibsoX0PPbPeJG/OuyPXEueY3aD1+Y+HE7GdBfjQsKO24qWJGN73Ilojot
INmfZuKHZtjiq/JZBr9YWmCA3Wul2XS1UpA7Gc7Nr/XNWRaT/vLLjFM95yHyLPCe
/S1qn4uR8pYbZfq4et/9kCyDpLiV58Cu2y/ZhXnHdlFmIAKTuWDY7sx3Hf98odCh
ua95F9irZzA2FbwsG50GoIWHxXiZN3tgAXZiqbU3NZXRVWURIyX3EPyL0vEsdeQk
3ZUzs1sFWBwKhMHJ1kghilfwsXUI0ML7x/eAnWypkrw4VBI4gCuG41qC02eA7qY/
juGxLcYCeEGkfecknmN40SPkHyr3wsZNtECYsslpHuxEwDhCWNPNVWznxhyfQFvf
d5d5eQtbVseQ/QoP3SqdxEk9YMb511G6UbOoXDiS9VPLXvL3GV9xKSp/tv7ZYVrV
vOdTjL7IE3oA9BAaRy2JX6r+09J1U2MQlWW/aoxogvU9r58Sz55TeKJ0GzEtsr4W
cCyizQQTCRMMNneNVcdP
=I/DW
-----END PGP SIGNATURE-----

**Comment by dilijev** [comment]
I agree, I don't want to rearchitect the program . A hack is acceptable if it lets us unit test things which don't truly depend on the GUI elements. I would recommend doing this by macros which would stand in for the actual privacy quantifiers. This will reduce the amount of "work" we have to make the switch, and keep the code in the actual classes cleaner.

#ifdef WX_MAXIMA_UNIT_TEST
#define wx_private public
#else
#define wx_private private
#endif

The macro definition code would have to be in a file which is directly or indirectly included in every compilation unit.

**Comment by gunterkoenigsmann** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

> The macro definition code would have to be in a file which is 
> directly or indirectly included in every compilation unit.
Or in the testbench.

As framework for the actual test I am used to use cppunit - which
basically does do little more than a few printf commands, but is alway
said to be standard. If I have time I can set up a few tests. But if
you know a better framework I wouldn't be too unhappy.
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1

iQJ8BAEBCgBmBQJU4Y3XXxSAAAAAAC4AKGlzc3Vlci1mcHJAbm90YXRpb25zLm9w
ZW5wZ3AuZmlmdGhob3JzZW1hbi5uZXQ1M0YwNDdDRTY2QjkxQjBGNzI0QzU0NUQ1
Qzg2QzBFNDIxMUQ1QjhFAAoJEFyGwOQhHVuO3c4P/0apS+tLgmebTekSDn9FX76X
ClENZmKyLvPufeJf+C8b9oFcor4KDbjI/xBj4Z2cD4hj11dDxAtrLkF1jdco0XNx
e+A+kLFacbOG0lOf4/+xhcZWcHXeDrn6CkBjLiTz/5sPwHIH9yeGCO3xNYK3zndz
wlCgB37shalyhzw7lvY24phx6g+vYevkTdjJJNidTLGwLtHjpEZoJn/1jv2NXVkZ
7kjAXyVmrA1jfOM2ltvSINP9YOEq91SRGUSpRr+iZ1U52Sc8/tgIMkI4zebR2w5H
MNdrhlO3y4HoMj77K2ip2kZsOMHwfiq6T1qIx3XiHVQ851SvoV51iavkO672jvYc
X/CbvpPq/DkV+4/s2Ma/3qcDDfOVyfUome+JHdvcgwmSdrV7TgVa5e/JufpHe0Wn
pStifyB26WkwxDivyNwMtrKNnCq8dUjoHOtZ7i3aRl6Aq4yrAXDX7Uye1Y45e4zF
zGrRyYgDi8MNCNxz/RvQ/iQYqSw10lhnj4QHpOgGqJCvXJKSWBJpuy0Q3Nucoi20
ojtmjP6YhItkFfq/ki2Awz3awmpHO4Dy1ukFrIeB3/sr4k/qneElVGX0ozRYG6hx
P2hD7VOjvq4wIhHPBpFfD3VQ2y+u4aj7LuR+iTjf1OJRZCfUzFMsZiA84uNNw0DR
zZ8fsviPpjaqxe0loPUF
=6k1z
-----END PGP SIGNATURE-----

**Comment by dilijev** [comment]
cppunit is what I've used most often. Sounds good to me.

**Comment by KubaO** [comment]
I've added a couple unit tests using Catch2, and those are integrated into the build system and CI. I consider this issue closed; individual unit tests should be added as needed.

To enable larger-scale regression tests, I also wrote a minimal interprocess communications event synthesizer, so that Maxima can send events (like `SELECTALL`, `CUT`, `PASTE`, etc.) back to wxMaxima. That way we can do simple UI tests to reproduce crashes. It should be easy to expand this mechanism further, and to add some support in `mathml.lisp` - currently I'm resorting to writing xml from plain `:lisp` statements. It works OK. This landed in 03b00ce5.

Generally speaking, since we already got Maxima as a "scripting language", it's not a long shot to use it to drive UI tests. At the moment it is one-directional, i.e. Maxima can send events to wxMaxima, but there's no way to ask wxMaxima for anything.
